core/src/main/scala/org/apache/spark/rdd/InputFileBlockHolder.scala

```diff
@@ -68,11 +71,17 @@ private[spark] object InputFileBlockHolder {
     require(filePath != null, "filePath cannot be null")
     require(startOffset >= 0, s"startOffset ($startOffset) cannot be negative")
     require(length >= 0, s"length ($length) cannot be negative")
-    inputBlock.set(new FileBlock(UTF8String.fromString(filePath), startOffset, length))
+    inputBlock.get().set(new FileBlock(UTF8String.fromString(filePath), startOffset, length))
   }
 
   /**
    * Clears the input file block to default value.
    */
   def unset(): Unit = inputBlock.remove()
+
+  /**
+   * Initializes thread local by explicitly getting the value. It triggers ThreadLocal's
+   * initialValue in the parent thread.
+   */
+  def initialize(): Unit = inputBlock.get()
```

Comment:
Do we need this? The parent thread should always create thread local variable before the child thread, right?

Reply:
Oh, no, sorry, before `get` is invoked, the thread local variable doesn't get initial value.

Reply:
We need this. Here was where I was stuck and confused for a while. We create the thread local but it doesn't call `initialValue`.

So, previously, here's what happens:

1. Input iterator that calls `InputFileBlockHolder.set` is executed in a child thread (in Python writer thread)
2. The child thread calls **`initialValue`** (not `childValue`) and sets new reference
3. After that, the parent thread reads it (say, `InputFileBlockHolder.getXXX`). Here, the parent threads calls **`initialValue`** again because `initialValue` was never called.

After the fix,

1. At the start of the task, the parent thread calls `initialValue`
2. After that, the child thread calls **`childValue`** (not `initialValue`); therefore, it points the same reference.
3. After that, the parent threads can access to the same reference.

I double checked those so if I am not like totally mistaken, should be correct ^.

Reply:
oh haha you were faster than me ..